src/gui/text/qfontdatabase.cpp

```diff
@@ -559,45 +551,15 @@ QtFontFamily *QFontDatabasePrivate::family(const QString &f, FamilyRequestFlags
 {
    QtFontFamily *fam = nullptr;
 
-   int low  = 0;
-   int high = count;
-   int pos  = count / 2;
-   int res  = 1;
-
-   if (count) {
-      while ((res = families[pos]->m_familyName.compare(f, Qt::CaseInsensitive)) && pos != low) {
-         if (res > 0) {
-            high = pos;
-         } else {
-            low = pos;
-         }
-         pos = (high + low) / 2;
-      }
-
-      if (! res) {
-         fam = families[pos];
-      }
-   }
-
-   if (!fam && (flags & EnsureCreated)) {
-      if (res < 0) {
-         pos++;
-      }
-
-      if (! (count % 8)) {
-         QtFontFamily **newFamilies = (QtFontFamily **)
-            realloc(families,
-               (((count + 8) >> 3) << 3) * sizeof(QtFontFamily *));
-         Q_CHECK_PTR(newFamilies);
-         families = newFamilies;
-      }
-
-      QtFontFamily *family = new QtFontFamily(f);
-      memmove(families + pos + 1, families + pos, (count - pos)*sizeof(QtFontFamily *));
-      families[pos] = family;
-      count++;
+   auto key = f.toLower();
 
-      fam = families[pos];
+   if (flags & EnsureCreated) {
+      auto res = families.insert(key, {f});
+      fam = &res.value();
+   } else {
+      auto res = families.find(key);
+      if(res != families.end())
+         fam = &res.value();
```

Comment:
It seems like it might be better to change the data type of `fam` to an iterator, rather than taking the address of an element in the container. It would be more clear, I think.

Reply:
I agree, but note that iterators can be invalidated, even when pointers/references into the unordered_map are not; not sure if it matters in this case, just though I'd point it out.

cppreference on unordered_map:
> References and pointers to either key or data stored in the container are only invalidated by erasing that element, even when the corresponding iterator is invalidated.

Reply:
Jan is right. And given other functions call `family` without exactly caring about the implementation details, returning a pointer is probably the right choice. 

Would this be better?

```c++
QtFontFamily *QFontDatabasePrivate::family(const QString &f, FamilyRequestFlags flags)
{
   QHash<QString, QtFontFamily>::iterator fam;

   auto key = f.toCaseFolded();

   if (flags & EnsureCreated) {
      fam = families.insert(key, {f});
   } else {
      fam = families.find(key);
   }

   if ((fam != families.end()) && (flags & EnsurePopulated)) {
      fam->ensurePopulated();
   }

   return (fam != families.end()) ? &fam.value() : nullptr;
}
```